src/main/java/net/minecraftforge/common/advancements/ICustomAdvancementReward.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.advancements;
+
+import com.google.gson.JsonObject;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+
+/**
+ * Interface for a custom advancement reward
+ */
+public interface ICustomAdvancementReward
+{
+
+    /**
+     * Called when the advancement is granted to the given player.
+     * Only runs on the server.
+     *
+     * @param player the player the advancement is granted to
+     */
+    void grant(ServerPlayer player);
+
+    /**
+     * Get the serializer for this custom advancement reward
+     * @return the serializer for this custom advancement reward
+     */
+    Serializer<?> getSerializer();
+
+    abstract class Serializer<T extends ICustomAdvancementReward> extends ForgeRegistryEntry<Serializer<?>>
+    {
+        public abstract T deserialize(JsonObject json);
+        public abstract JsonObject serialize(T reward);
+    }
```

Comment:
While I do agree with the concept, I think we should be moving towards codecs for these types of solutions. Though, this is purely an opinion of mine.

Reply:
I would prefer Codecs too but the problem is a Codec can't be registered in a Registry.

Reply:
True, but you can wrap the codec in another object. Since there is an RFC to remove the `IForgeRegistryEntry` limitation in the future, I would say that would cover that issue and provide for a swift transition out of the wrapping object.

Reply:
So I should just convert it to a class holding a codec

Reply:
That's my opinion on the matter on it yeah. Not sure if anyone else has a differing opinion though, so it may be worthwhile to hold off on it for a bit until someone else chimes in on the idea.

Reply:
The status quo is that these kinds of customizable data-driven objects have serializers which are registered with an existing Forge registry. While it may be worth looking into the matter for possible changes across all similar objects in the future, I'm for keeping the status quo on this for now.